# Homework 1 Code and Report
AUFAR Rezka and LASBLEIS Alexandre 

### Step 1 : Simulate Rankings of Relevance for E and P (5 points)

In the first step you will generate pair s of rankings of relevance, for the production P and experimental E,
respectively, for a hypothetical query q . Assume a 3-graded relevance, i.e. {N, R, HR}. Construct all
possible P and E ranking pairs of length 5, for which E outperforms P.

Example:<br>
P: {N N N N N}<br>
E: {N N N N R}<br>
… <br>
P: {HR HR HR HR R}<br>
E: {HR HR HR HR HR}<br>

(Note 1: If you do not have enough computational power, sample 1000 pair uniformly at random to show
your work.)
##### Code

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand
from scipy.stats import bernoulli

In [2]:
#convert int to relevance
#_list : list of integer to be converted
#return : list of string where every integer has been replaced by its relevance interpretation
def to_string(_list): 
    L=[]
    for i in _list:
        if i==0:
            L.append("N ")
        elif i==1:
            L.append("R ")
        elif i==2:
            L.append("HR")
        else: 
            L.append("? ")
    return L

In [3]:
pairs = [] #variable used to store all pairs 
Ps = []
Es = []

#Create all combinaison for P, there are 3^5 = 243 possibilities since P is of length 5 and each member have 3 possible values:
# 0 (representing "N" ), 1 (representing "R), 2 (representing "HR)
for p_possibility in range(243): 
    P=[None]*5
    #create all five values for P 
    for p_pos in range(5):
        P[p_pos]= p_possibility // (np.power(3,p_pos)) %3 #formula to create all combinaison
    #we do the same for E    
    for e_possibility in range(243): 
        E=[None]*5
        for e_pos in range(5):
            E[e_pos]= e_possibility // (np.power(3,e_pos)) %3 #formula to create all combinaison
        pairs.append((P,E)) #adding the pair since every pair is valid
        #Memorize all created Es but only once
        if p_possibility==0:
            Es.append(E)
        #memorize all created Ps
    Ps.append(P)

In [4]:
#TESTING
print(len(pairs)) #= 59049 which is equal to the expected length = 3 ^10 = 59049
#print(pairs) #print all pairs but this takes a lot of place

#print a random pair to show that it makes sense 
r = rand.randint(0,59049-1)
print("P: " , to_string(pairs[r][0]), "E: ",to_string(pairs[r][1]))

#check if Es is as expected 
print(len(Es))#expect 243 = 3^5 
print(Es)

59049
('P: ', ['N ', 'R ', 'R ', 'R ', 'R '], 'E: ', ['N ', 'N ', 'R ', 'N ', 'HR'])
243
[[0, 0, 0, 0, 0], [1, 0, 0, 0, 0], [2, 0, 0, 0, 0], [0, 1, 0, 0, 0], [1, 1, 0, 0, 0], [2, 1, 0, 0, 0], [0, 2, 0, 0, 0], [1, 2, 0, 0, 0], [2, 2, 0, 0, 0], [0, 0, 1, 0, 0], [1, 0, 1, 0, 0], [2, 0, 1, 0, 0], [0, 1, 1, 0, 0], [1, 1, 1, 0, 0], [2, 1, 1, 0, 0], [0, 2, 1, 0, 0], [1, 2, 1, 0, 0], [2, 2, 1, 0, 0], [0, 0, 2, 0, 0], [1, 0, 2, 0, 0], [2, 0, 2, 0, 0], [0, 1, 2, 0, 0], [1, 1, 2, 0, 0], [2, 1, 2, 0, 0], [0, 2, 2, 0, 0], [1, 2, 2, 0, 0], [2, 2, 2, 0, 0], [0, 0, 0, 1, 0], [1, 0, 0, 1, 0], [2, 0, 0, 1, 0], [0, 1, 0, 1, 0], [1, 1, 0, 1, 0], [2, 1, 0, 1, 0], [0, 2, 0, 1, 0], [1, 2, 0, 1, 0], [2, 2, 0, 1, 0], [0, 0, 1, 1, 0], [1, 0, 1, 1, 0], [2, 0, 1, 1, 0], [0, 1, 1, 1, 0], [1, 1, 1, 1, 0], [2, 1, 1, 1, 0], [0, 2, 1, 1, 0], [1, 2, 1, 1, 0], [2, 2, 1, 1, 0], [0, 0, 2, 1, 0], [1, 0, 2, 1, 0], [2, 0, 2, 1, 0], [0, 1, 2, 1, 0], [1, 1, 2, 1, 0], [2, 1, 2, 1, 0], [0, 2, 2, 1, 0], [1, 2, 2, 1, 0], [2, 2, 2,

##### Explanations
In order to contruct all possible pairs of P and E with a 3-graded relevance {N, R, HR}, we first encode those relevance as integers: <br>
* 0 for N <br>
* 1 for R <br>
* 2 for HR <br>
This encoding allows us to do easy comparison between elements.
Then we created all pairs using this formula : $value = \frac{possibility}{3^{pos}} [3]$ with $possibility \in \{0,...,243\}$ and $ pos \in \{0,1,2,3,4\}$.

This give the following sequence depending on the position of the element in the list: <br>
* 0: $\{0,1,2,0,1,2,0,1,2,0,1,2...\}$
* 1: $\{0,0,0,1,1,1,2,2,2,0,0,0...\}$
* 2: $\{0,0,0,0,0,0,0,0,0,1,1,1,...\}$
* ... and so on

This formula is then used for the production P and the experiment E. 

##### Analysis
According to the formula we showed before, we should find all the possible combination. Futhermore, the number of pairs found is equal to the one we expected $3^{10}$ and the number of E and P are $3^{5}$ each which is also what is expected. 


### Step 2 : Implement Evaluation Measures (15 points)

Implement 1 binary and 2 multi-graded evaluation measures out of the 7 measures mentioned above.

(Note 2: Some of the aforementioned measures require the total number of relevant and highly relevant
documents in the entire collection – pay extra attention on how to find this)

##### code

In [5]:
#TODO: Remove for evaluation 

#Recall at rank k 
def Binary_measure_recal(test_set, r):
    num_test_rel = 0 
    total_num_rel=0
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_test_rel+=1
            total_num_rel+=1
    return float(num_test_rel) / total_num_rel

#Average precision at rank k 
def Binary_measure_avg_precision(test_set, r):
    num_rel=0
    sum_precision = 0 
    total_num_rel=0
    temp="("
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_rel+=1
                temp+=" "+str(num_rel)+"/"+str(i+1)+"+"
                sum_precision+=float(num_rel)/(i+1)
            total_num_rel+=1
    temp=temp[:len(temp)-1]+")"
    #print(temp,"/",total_num_rel)
    return float(sum_precision) / total_num_rel
#Normalized discount cumulative gain nDCG at rank r
def nDCG(test_set, r):
    orderedSet= list(test_set)
    orderedSet.sort(reverse=True)
    return DCG(test_set, r) / DCG(orderedSet, r)

In [6]:
#BINARY 
#Precision at rank r 
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the precision 
def Binary_measure_precision(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    num_test_rel = 0 
    for i in range(r):
        #According to our notation, if test_set[i]>0, the document is relevant
        if test_set[i]>0:
            num_test_rel+=1
    return float(num_test_rel) / r #precision is number of relevant doc/ number of doc considered

#MULTI GRADED 
#Discount cumulative gain DCG at rank r
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the gain 
def DCG(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    res=0
    for k in range(1,r+1):#r is the rank so it starts at 1
        rel_r = test_set[k-1]#index of element = rank -1 since rank starts at 1 and index at 0
        res+=float(np.power(2,rel_r)-1)/(np.log2(1+k))
    return res

#Rank Biased Precision w ith persistence parameter 𝜃 =0.8
#test_set: the collection to be tested 
#theta: the theta parameter of the RBP formula
#return : the precision
def RBP (test_set,theta=0.8):
    res=0
    for k in range(1,len(test_set)+1):
        rel_k=test_set[k-1]
        res+= rel_k * np.power(theta,k-1) * (1.0-theta)
    return res
        

In [7]:
#TESTING
test=        [1,0,2,5,0,1,0,0,0,1]#length 10
test_perfect=[5,2,1,1,1,0,0,0,0,0]
test_worst=  [0,0,0,0,0,1,1,1,2,5]

#The following results are rounded in order to make them readable
#Specific ordering 
print("Specific ordering")
for r in range(1,len(test)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test,r),3),'DCG', round(DCG(test,r),3),
          'RBP', round(RBP(test),3))

#Perfect ordering 
print("\nPerfect ordering")
for r in range(1,len(test_perfect)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test_perfect,r),3),'DCG', round(DCG(test_perfect,r),3),
          'RBP', round(RBP(test_perfect),3))

#Worst ordering 
print("\nWorst ordering")
for r in range(1,len(test_worst)+1):
    print('Rank = ' , r,'Precision',round(Binary_measure_precision(test_worst,r),3),'DCG', round(DCG(test_worst,r),3),
          'RBP', round(RBP(test_worst),3))

Specific ordering
('Rank = ', 1, 'Precision', 1.0, 'DCG', 1.0, 'RBP', 1.06)
('Rank = ', 2, 'Precision', 0.5, 'DCG', 1.0, 'RBP', 1.06)
('Rank = ', 3, 'Precision', 0.667, 'DCG', 2.5, 'RBP', 1.06)
('Rank = ', 4, 'Precision', 0.75, 'DCG', 15.851, 'RBP', 1.06)
('Rank = ', 5, 'Precision', 0.6, 'DCG', 15.851, 'RBP', 1.06)
('Rank = ', 6, 'Precision', 0.667, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 7, 'Precision', 0.571, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 8, 'Precision', 0.5, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 9, 'Precision', 0.444, 'DCG', 16.207, 'RBP', 1.06)
('Rank = ', 10, 'Precision', 0.5, 'DCG', 16.496, 'RBP', 1.06)

Perfect ordering
('Rank = ', 1, 'Precision', 1.0, 'DCG', 31.0, 'RBP', 1.632)
('Rank = ', 2, 'Precision', 1.0, 'DCG', 32.893, 'RBP', 1.632)
('Rank = ', 3, 'Precision', 1.0, 'DCG', 33.393, 'RBP', 1.632)
('Rank = ', 4, 'Precision', 1.0, 'DCG', 33.823, 'RBP', 1.632)
('Rank = ', 5, 'Precision', 1.0, 'DCG', 34.21, 'RBP', 1.632)
('Rank = ', 6, 'Precision', 0.833, 'DCG', 34.2

##### Explanations
For Precision and DCG, we need to ensure that rank k does not exceed the documents length, and k cannot be zero.<br>
For DCG and RBP, we also decided that the value of the relevant would be equal to the value at that rank in the list. For example if an example is extremely relevant, he might get a value of 5 (as in the example) which leads to higher gain in DCG than a document which is only highly relevant (value of 2).<br>
Otherwise our implementation follows the definition of the method. 

##### Analysis
The result for Precision is highly dependant on the number of non-relevant in the cut-off collection. As k increases, the more non-relevant exists in the cut-off, the lower the precision will be. If we consider a rank equal to the length of the list, the precision is the same whatever the order is. Precision does not that into account how good the relevance of a document is, it only cares if a document is relevant or not. 

For DCG, the result depends on the degree of the relevance and the order of it. If the highest relevant score exists in the first position of the cut-off, the higher DCG will be. We showed that by displaying the DCG score with perfect ordering which is higher at any rank than the "specific ordering". Those value may be hard to use to compare different collections due to the fact that it is sensitive to the degree of relevance. A way to cope with that problem is to use its alternative version: the nDCG (normalized version). 

For RBP, we disregard rank and therefore we have the same result no matter the rank is. This measure pays great attention to the ordering as shown above: the best score is obtained with a perfect ordering whereas the worst one is obtained with the worst ordering. The model behind RBP considers that when you see a document, you have a probability to see the next document otherwise you stop. This model can be improved by taking into account how relevant is the document that we are looking at. Depending on this relevance, there is a probability to see the next document or to stop. This second model is called ERR model.



### Step 3 : Calculate the measure (5 points)

For the three measures and all P and E ranking pairs constructed above calculate the difference: 𝛥measure
= measure E -measure P . Consider only those pairs for which E outperforms P.

##### code

In [8]:
#Evaluate if E outperfoms P with respect to a specific measure. 
#methond_name: the measure used to test. It can be "precision", "DCG" or "RBP"
#E: 
#P:
#rank: 
#return : (outperforms, delta measure) where outperforms is True if E outperfors P with respect to the measure method for the
#rank and delta measure is the difference between to measure of E and P for the measure method.
def outperforms(methond_name,E,P,rank):
    if methond_name=="precision":
        delta = Binary_measure_precision(E,rank)-Binary_measure_precision(P,rank)
        return (delta>0), delta 
    elif methond_name=="DCG":
        delta=DCG(E,rank)-DCG(P,rank)
        return (delta>0), delta 
    elif methond_name=="RBP":
        delta=RBP(E)-RBP(P)
        return (delta>0), delta 

#variable for precision measure 
outperf_pairs_prec=[] #memorize pairs for which E outperfoms P
pairs_set_prec = [] 
d_measure_prec=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

#variable for DCG measure 
outperf_pairs_DCG=[] #memorize pairs for which E outperfoms P
pairs_set_DCG = []
d_measure_DCG=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

#variable for RBP measure
outperf_pairs_RBP=[] #memorize pairs for which E outperfoms P
pairs_set_RBP = []
d_measure_RBP=[] #memorize "measure(E)-measure(P)" for which E outperfoms P

rank = 4
for i in range(len(pairs)):#pairs made of (P,E)
    #test if E outperforms P with respect to a specific method
    outperforms_prec,delta_prec= outperforms("precision",pairs[i][1],pairs[i][0],rank)
    outperforms_DCG,delta_DCG= outperforms("DCG",pairs[i][1],pairs[i][0],rank)
    outperforms_RBP,delta_RBP= outperforms("RBP",pairs[i][1],pairs[i][0],rank)

    #Memorize some "precision" information only if E outperfom P
    if(outperforms_prec):
        outperf_pairs_prec.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_prec.append(''.join(str(x) for x in temp))
        d_measure_prec.append(delta_prec)
    
    #Memorize some "DCG" information only if E outperfom P
    if(outperforms_DCG):
        outperf_pairs_DCG.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_DCG.append(''.join(str(x) for x in temp))
        d_measure_DCG.append(delta_DCG)
    
    #Memorize some "RBP" information only if E outperfom P
    if(outperforms_RBP):
        outperf_pairs_RBP.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_RBP.append(''.join(str(x) for x in temp))
        d_measure_RBP.append(delta_RBP)     


In [11]:
#TESTING

print ("Number of pairs for which E outperforms P for all three measures: ",len(set.intersection(set(pairs_set_prec), set(pairs_set_DCG), set(pairs_set_RBP))))
print ("Pair in precision and in DCG:",rand.sample(set.intersection(set(pairs_set_prec), set(pairs_set_DCG)),1))

# pick result that is in prec but is not in DCG
print ("Pair in precision and not in DCG:",rand.sample(set(pairs_set_prec) - set(pairs_set_DCG),1))
# pick result that is in DCG but is not in prec
print ("Pair in DCG and not in precision:",rand.sample(set(pairs_set_DCG) - set(pairs_set_prec),1))


Number of pairs for which E outperforms P for all three measures:  17184
Pair in precision and in DCG: ['1102100122']
Pair in precision and not in DCG: ['1111222102']
Pair in DCG and not in precision: ['2222212111']


In [9]:
pair_prec_not_DCG = [[1,1,1,1,2],[2,2,1,0,2]] # pair E, P 
pair_DCG_not_prec = [[2,2,2,2,2],[1,2,1,1,1]] # pair E, P 
print("Pair in precision and not in DCG:",pair_prec_not_DCG, "precision E:",
      Binary_measure_precision(pair_prec_not_DCG[0],rank),"precision P:",
      Binary_measure_precision(pair_prec_not_DCG[1],rank),"DCG E: ",
     DCG(pair_prec_not_DCG[0],rank),"DCG P: ",
     DCG(pair_prec_not_DCG[1],rank) )

print("Pair in DCG and not in precision:",pair_DCG_not_prec, "DCG E: ",
     DCG(pair_DCG_not_prec[0],rank),"DCG P: ",
     DCG(pair_DCG_not_prec[1],rank),"precision E:",
      Binary_measure_precision(pair_DCG_not_prec[0],rank),"precision P:",
      Binary_measure_precision(pair_DCG_not_prec[1],rank))

('Pair in precision and not in DCG:', [[1, 1, 1, 1, 2], [2, 2, 1, 0, 2]], 'precision E:', 1.0, 'precision P:', 0.75, 'DCG E: ', 2.5616063116448506, 'DCG P: ', 5.3927892607143724)
('Pair in DCG and not in precision:', [[2, 2, 2, 2, 2], [1, 2, 1, 1, 1]], 'DCG E: ', 7.6848189349345519, 'DCG P: ', 3.8234658187877653, 'precision E:', 1.0, 'precision P:', 1.0)


##### Explanations
For precision, DCG, and RBP, we calculate the difference between all combination of E and P that we have generated in Step 1, then we append every difference that is greater than zero, which means that E outperforms P.
##### Analysis
For each three measures, we have different pairs in which E outperforms P. There are 17184 pairs where E outperforms P in all three measures. However, there exists some pairs where E outperforms P for one kind of measure, but not the other. <br> 
For example, E=[1,1,1,1,2] outperforms P=[2,2,1,0,2] in precision (delta of 0.25), but not in DCG (delta of -2.83118294971) . This is due to the fact that precision does not check order and the degree of relevance. The same happens the opposite way : E=[2,2,2,2,2] outperforms P=[1,2,1,1,1] in DCG (delta of 3.86135311614), but not in precision (delta of 0). <br>
This indicates that in real world application, it is important to implement different kinds of measure so we will have broader feedback to optimize the algorithm.

## Step 4 : Implement Interleave (15 points)

Implement 2 interleaving algorithms: (1) Team-Draft Interleaving OR Balanced Interleaving, AND (2),
Probabilistic Interleaving. The interleaving algorithms should (a) given two rankings of relevance
interleave them into a single ranking, and (b) given the users clicks on the interleaved ranking assign
credit to the algorithms that produced the rankings.

##### code

In [10]:
# assignment assumption : E and P always return different documents
# general parameter : consider two pairs can contain same documents 

def team_draft_interleave(ranking_pairs, general=False, clicks=[]):
    # interleave two list
    interleaved_list = []
    added_list = []
    assert len(ranking_pairs[0]) ==  len(ranking_pairs[1])
    for i, elem in enumerate(ranking_pairs[0]):
        coin = rand.choice([0,1])
        if(coin == 0):
            if(general and ranking_pairs[0][i] not in added_list):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                added_list.append(ranking_pairs[0][i])
            if(general and ranking_pairs[1][i] not in added_list):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                added_list.append(ranking_pairs[1][i])
            if(not general):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                interleaved_list.append((ranking_pairs[1][i], "P"))
        if(coin == 1):
            if(general and ranking_pairs[1][i] not in added_list):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                added_list.append(ranking_pairs[1][i])
            if(general and ranking_pairs[0][i] not in added_list):
                interleaved_list.append((ranking_pairs[0][i], "E"))
                added_list.append(ranking_pairs[0][i])
            if(not general):
                interleaved_list.append((ranking_pairs[1][i], "P"))
                interleaved_list.append((ranking_pairs[0][i], "E"))
    # generate user clicks randomly
    if(not general):
        user_click = [rand.randrange(0, 2) for _ in range(0, len(interleaved_list))]
    else:
        user_click = clicks
    # calculate scores
    E_wins, P_wins = 0, 0
    for i, click in enumerate(user_click):
        if click == 1:
            if interleaved_list[i][1] == "E":
                E_wins += 1
            if interleaved_list[i][1] == "P":
                P_wins += 1
    return interleaved_list, user_click, E_wins, P_wins

In [11]:
r = rand.randint(0,59049-1)
interleaved_list, user_click, E_wins, P_wins = team_draft_interleave(pairs[r])
print("Interleaved list ",interleaved_list,"user click" ,user_click,"number of E wins" ,E_wins,"number of P wins" ,P_wins )

('Interleaved list ', [(2, 'P'), (0, 'E'), (1, 'E'), (1, 'P'), (0, 'E'), (2, 'P'), (0, 'E'), (2, 'P'), (2, 'E'), (2, 'P')], 'user click', [0, 0, 1, 0, 1, 1, 1, 1, 1, 1], 'number of E wins', 4, 'number of P wins', 3)


In [12]:
# general test
general_test = ([1,2,3,4,5], [6,7,8,1,2])
team_draft_interleave(general_test, True, [1,0,0,1,0])

([(6, 'P'),
  (1, 'E'),
  (2, 'E'),
  (7, 'P'),
  (8, 'P'),
  (3, 'E'),
  (4, 'E'),
  (5, 'E')],
 [1, 0, 0, 1, 0],
 0,
 2)

In [28]:
# tested the result with existing library (https://github.com/mpkato/interleaving/tree/master/interleaving)

import interleaving
a = [1, 2, 3, 4, 5] # Ranking 1
b = [6, 7, 8, 9, 10] # Ranking 2
method = interleaving.Probabilistic([a, b]) # initialize an interleaving method
ranking = method.interleave() # interleaving
print ranking

clicks = [0, 3] # observed clicks, i.e. documents 1 and 2 are clicked
result = interleaving.Probabilistic.evaluate(ranking, clicks)
print result # (0, 1) indicates Ranking 1 won Ranking 2.

[6, 1, 2, 4, 10]
[]


In [32]:
import copy

def softmax(tau, length):
    total_denominator = 0
    softmax_distribution = []
    for i in range(1, length+1):
        total_denominator += float(1.0/np.power(i,tau))
    for i in range(1, length+1):
        softmax_distribution.append(float(1.0/np.power(i,tau) / (total_denominator)))
    return softmax_distribution

def probabilistic_interleave(ranking_pairs):
    # generate softmax probability distribution with tau = 3 (Hofmann, 2011)
    assert len(ranking_pairs[0]) ==  len(ranking_pairs[1])
    E = copy.copy(ranking_pairs[0])
    P = copy.copy(ranking_pairs[1])
    softmax_distribution_E = softmax(3,len(E))
    softmax_distribution_P = softmax(3,len(P))
    # interleave two list with softmax probability distribution
    interleaved_list = []
    for _ in range(len(E) + len(P)):
        coin = rand.choice([0,1])
        if(coin == 0):
            # check if E is empty then just sample from P
            if not E:
                # choose element from P randomly
                picked = np.random.choice(P, 1, p=softmax_distribution_P)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "P"))
                # remove the picked element
                P.remove(picked)
                # renormalize the distribution
                softmax_distribution_P = softmax(3,len(P))
            # else sample from E
            else:
                # choose element from E randomly
                picked = np.random.choice(E, 1, p=softmax_distribution_E)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "E"))
                # remove the picked element
                E.remove(picked)
                # renormalize the distribution
                softmax_distribution_E = softmax(3,len(E))
        if(coin == 1):
            # check if P is empty then just sample from E
            if not P:
                # choose element from E randomly
                picked = np.random.choice(E, 1, p=softmax_distribution_E)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "E"))
                # remove the picked element
                E.remove(picked)
                # renormalize the distribution
                softmax_distribution_E = softmax(3,len(E))
            # else sample from P
            else:
                # choose element from P randomly
                picked = np.random.choice(P, 1, p=softmax_distribution_P)[0]
                # append it to interleaved list
                interleaved_list.append((picked, "P"))
                # remove the picked element
                P.remove(picked)
                # renormalize the distribution
                softmax_distribution_P = softmax(3,len(P))
    # generate user clicks randomly
    user_click = [rand.randrange(0, 2) for _ in range(0, len(interleaved_list))]
    # calculate scores
    E_wins, P_wins = 0, 0
    for i, click in enumerate(user_click):
        if click == 1:
            if interleaved_list[i][1] == "E":
                E_wins += 1
            if interleaved_list[i][1] == "P":
                P_wins += 1
    return interleaved_list, user_click, E_wins, P_wins

In [33]:
r = rand.randint(0,59049-1)
#team_draft_interleave(pairs[r])

In [36]:
probabilistic_interleave(pairs[r])

([(2, 'P'),
  (2, 'E'),
  (2, 'P'),
  (1, 'P'),
  (1, 'E'),
  (0, 'P'),
  (0, 'P'),
  (1, 'E'),
  (1, 'E'),
  (1, 'E')],
 [0, 0, 1, 1, 0, 0, 0, 1, 0, 1],
 2,
 2)

To do probabilistic interleave : 
- Make it general

##### Explanations

##### Analysis


## Step 5 : Implement User Clicks Simulation ( 25 points)

Having interleaved all the ranking pairs an online experiment could be ran. However, given that we do not
have any users (and the entire homework is a big simulation) we will simulate user clicks.<br>
We have considered a number of click models including:<br>
* 1. Random Click Model (RCM)
* 2. Position-Based Model (PBM)
* 3. Simple Dependent Click Model (SDCM)
* 4. Simple Dynamic Bayesian Network (SDBN)

Consider two different click models, (a) the Random Click Model (RCM), and (b) one out of the
remaining 3 aforementioned models. The parameters of some of these models can be estimated using the
Maximum Likelihood Estimation (MLE) method, while others require using the
Expectation-Maximization (EM) method. Implement the two models so that:
* (a) there is a method that learns the parameters of the model given a set of training data, 
* (b) there is a method that predicts the click probability given a ranked list of relevance labels, 
* (c) there is a method that decides - stochastically -whether a document is clicked based on these probabilities.<br>

Having implemented the three click models, estimate the model parameters using the Yandex Click Log

##### code

In [8]:
#Random Click Model (RCM) 
def learns_param_RCM(training_data):
    numb_doc = 0
    num_click=0
    for data in training_data:
        numb_doc+=1
        if("C" in data[2]):
            num_click+=1
    return float(num_click)/numb_doc  

def predict_click_proba_RCM(training_data,relevance_labels) :
    rau = learns_param_RCM(training_data)
    #since the probability is rau for every document, we return a list of rau that has the length of the relevance labels list
    return [rau]*len(relevance_labels)

def is_doc_clicked_RCM(training_data,relevance_labels,document):
    proba_distrib= predict_click_proba_RCM(training_data,relevance_labels)
    return (bernoulli.rvs(proba_distrib[document])==1)

In [14]:
#open file and preprocess data
training_data=[]
relevance_labels= [1,0,2,5,0,1,1,2,5]
document= 3

f= open("YandexRelPredChallenge.txt","r")
for line in f:
    training_data.append(line.split("\n")[0].split('\t'))
f.close()

print("isDocClicked", is_doc_clicked_RCM(training_data,relevance_labels,document))

isDocClicked False


##### Explanations

##### Analysis